feat(tutorial): add k8s storage tuto to accompany video (#2669)
scaleway/docs-content: tutorials/get-started-kubernetes-storage/index.mdx

categories:
  - kubernetes
dates:
  validation: 2024-01-29
  posted: 2024-01-29
---

This tutorial accompanies the fourth [video demonstration](https://www.youtube.com/watch?v=OAR_3Fuwtw0) in our series to help users get started with Kubernetes. We walk you through Kubernetes fundamentals for beginners. In this installment, we show you how to add storage to your [Scaleway Kubernetes Kapsule](https://www.scaleway.com/en/kubernetes-kapsule/).  

First, we address the question of why we might need to add storage to our cluster, comparing **persistent** storage with **system** storage, and digging into **stateless** vs. **stateful** applications. We then show how to create a stateful application in a cluster via a **StatefulSet object**, which provisions persistent storage via a PersistentVolumeClaim, which in turns provisions and manages storage using the Scaleway Block Storage product.

Previous videos in this series covered the basics of [containers and Docker](/tutorials/get-started-containers-docker/), [deploying an app with Kapsule](/tutorials/get-started-deploy-kapsule/), and [adding a Load Balancer](/tutorials/get-started-kubernetes-loadbalancer/).

> <Macro id="iam-requirements" />

Reviewer: You might need to rebase and rewrite the requirements

Reviewer 2: @RoRoJ I updated the requirements part (see below), not sure if it needs a rebase as this is a new page

> <Macro id="iam-requirements" />
>  
> <Message type="requirement">
> 
>   - You have an account and are logged into the [Scaleway console](https://console.scaleway.com)
>   - You have [generated your API key](/identity-and-access-management/iam/how-to/create-api-keys/)
>   - You have [created a Kubernetes Kapsule](/containers/kubernetes/how-to/create-cluster/) and [configured kubectl](/containers/kubernetes/how-to/connect-cluster-kubectl/).
> 
> </Message>

Reviewer 2: ```suggestion
<Macro id="requirements" />

  - A Scaleway account logged into the [console](https://console.scaleway.com)
  - A valid [API key](/identity-and-access-management/iam/how-to/create-api-keys/)
  - A [Kubernetes Kapsule](/containers/kubernetes/how-to/create-cluster/) and [kubectl configured](/containers/kubernetes/how-to/connect-cluster-kubectl/).
```

## Fundamental storage concepts

### System volume vs persistent volumes

It is important to differentiate between Kuberenetes **system volumes** and **persistent volumes**. This tutorial is concerned with **persistent volumes**, which can be provisioned and then used by applications inside the cluster to store data. Kuberenetes **system volumes** are created automatically but can only be used by the Kubernetes system itself to store essential files. This is summed up in the table below:

| System volume                            | Persistent volume                          |
| ---------------------------------------- | ------------------------------------------ |
| For the k8s system itself                |  For applications in the cluster           |
| Ephemeral (lives and dies with the node) | Persistent (beyond the life of a pod/node) |
| Auto-created                             | Must be provisioned                        |

> You can view your cluster's system volumes in the **Instances > Volumes** section of the [console](https://console.scaleway.com/), and see how they are linked to the Instances that make up the nodes of your cluster. However, you shouldn't attempt to use or manage these volumes - leave it to Kubernetes Kapsule. 

Reviewer: ```suggestion
You can view your cluster's system volumes in the **Instances > Volumes** section of the [console](https://console.scaleway.com/), and see how they are linked to the Instances that make up the nodes of your cluster. However, you should not attempt to use or manage these volumes - leave it to Kubernetes Kapsule. 
```

| Each request is independent and doesn't use previous data   | Need to retrieve stored data                                         |
| Examples: microservice, [whoami](/tutorials/get-started-containers-docker/#example-2-a-more-complex-app), search engine request     | Examples: most modern applications (online banking, online shopping) |
| "Original" Kubernetes                                       | Kubernetes has now adapted to incorporate statefulness               |
| Do not need to provision storage                            | Must provision storage                                               |

## How do we provision storage?

### Provisioning Block Storage volumes

Cloud providers such as Scaleway provide and default **StorageClasses** for their managed Kubernetes products. StorageClasses define the classes of storage the provider can offer for the cluster. The user can list the StorageClasses available to by using the command `kubectl get storageclass`.

Scaleway provides [Block Storage](/storage/block/quickstart/) as its StorageClass, meaning that we offer persistent storage for Kapsule clusters via our Block Storage product. 

<Message type="note">

> Scaleway has upgraded the Block Storage used by our StorageClass, offering better performance. Our previous class was `scw-bss`, which has now been replaced by `sbs`, giving you faster read/write operations with lower latency.

Reviewer 3: typo, it's `scw-bssd`
now replaced by `sbs-default` to my default sbs only doesn't exist

> Scaleway has upgraded the Block Storage used by our StorageClass, offering better performance. Our previous class was `scw-bss`, which has now been replaced by `sbs`, giving you faster read/write operations with lower latency.

Author: ```suggestion
Scaleway has upgraded the Block Storage used by our StorageClass, offering better performance. Our previous class was `scw-bssd`, which has now been replaced by `sbs-default`, giving you faster read/write operations with lower latency.
```

</Message>

> Cloud providers also provide a **C**ontainer **S**torage **I**nterface (CSI) for their managed clusters. This is a standardized interface allowing communication between Kubernetes and the cloud provider's own storage products. CSI and StorageClasses make **dynamic volume provisioning** possible, whereby Kubernetes Kapsule users can create storage volumes for their clusters on-demand. They simply describe the storage required by referring to a StorageClass, and this is provisioned from Scaleway by the CSI.

Author: ```suggestion
Cloud providers also provide a **C**ontainer **S**torage **I**nterface (CSI) for their managed clusters. This is a standardized interface allowing communication between Kubernetes and the cloud provider's own storage products. CSI and StorageClasses make **dynamic volume provisioning** possible, whereby Kubernetes Kapsule users can create storage volumes for their clusters on-demand. They simply describe the storage required by referring to a StorageClass, and this is provisioned from Scaleway by the [CSI](https://github.com/scaleway/scaleway-csi).
```

### Kubernetes objects

> In order to provision persistent Block Storage resources in this tutorial, we will use a **PersistentVolumeClaim**. This is a Kubernetes object, which requests and claims a **PersistentVolume** resource. The PersistentVolumeClaim object requests a specific size, access mode and StorageClass for the PersistentVolume, meaning we can describe what we want without needing to worry about how it's provisioned.

Reviewer: ```suggestion
In order to provision persistent Block Storage resources in this tutorial, we will use a **PersistentVolumeClaim**. This is a Kubernetes object, which requests and claims a **PersistentVolume** resource. The PersistentVolumeClaim object requests a specific size, access mode and StorageClass for the PersistentVolume, meaning we can describe what we want without needing to worry about how it is provisioned.
```

<Lightbox src="scaleway-persistent-volume-claim.webp" alt ="" />

In our [previous tutorial](/tutorials/get-started-deploy-kapsule/) we used a **Deployment** object to deploy a stateless application. In this tutorial, we will use a **StatefulSet** object to deploy a stateful application.

> Both Deloyments and StatefulSets manage the deployment and scaling of a set of pods with identical replications of a containerized application. But StatefulSets can also provision and manage persistent storage for thier pods, via PersistentVolumeClaims. What's more, in order to ensure data consistency and synchronicity as pods read and write to storage, the pods of a StatefulSet have sticky, persistent identities as shown below:

Reviewer: ```suggestion
Both Deployments and StatefulSets manage the deployment and scaling of a set of pods with identical replications of a containerized application. But StatefulSets can also provision and manage persistent storage for their pods, via PersistentVolumeClaims. What's more, in order to ensure data consistency and synchronicity as pods read and write to storage, the pods of a StatefulSet have sticky, persistent identities as shown below:
```

    kubectl get all
    ```

2. Create a new file called ss.yaml:

    ```
    nano lb.yaml
    ```

    This file will be a manifest for our StatefulSet object. It will describe the stateful application we want to deploy in the cluster, and the persistent storage volumes that are required for the application.

3. Copy and paste the following text into the file:

    ```yaml
    apiVersion: apps/v1
    kind: StatefulSet
    metadata:
      name: ss-csi-scw
    spec:
      serviceName: ss-csi-scw
      replicas: 5
      selector:
        matchLabels:
          app: ss-csi-scw
          provider: scaleway
      template:
        metadata:
          labels:
            app: ss-csi-scw
            provider: scaleway
        spec:
          containers:
          - name: busy-pvc-scw
            image: busybox
            volumeMounts:
            - mountPath: "/data"
              name: csi-vol-scw
            command: ["sleep", "3600"]
      volumeClaimTemplates:
      - metadata:
          name: csi-vol-scw
        spec:
          accessModes: ["ReadWriteOnce"]
          storageClassName: sbs-default
          resources:
            requests:
              storage: 1Gi
    ```

>     Essentially, we are creating a **StatefulSet** object, with **5 replica pods**, running a lightweight container called **busyboxùù that will just sleep int eh background for an hour. The StatefulSet will create a **PersistentVolumeClaim** for each pod, creating a 1GB Scaleway **Block Storage volume** for each pod, mounted at **/data**. For a full explanation of the various fields in the yaml manifest, see below:

Reviewer: ```suggestion
    Essentially, we are creating a **StatefulSet** object, with **5 replica pods**, running a lightweight container called **busybox** that will just sleep in the background for an hour. The StatefulSet will create a **PersistentVolumeClaim** for each pod, creating a 1GB Scaleway **Block Storage volume** for each pod, mounted at **/data**. For a full explanation of the various fields in the yaml manifest, see below:
```

          - **storageClassName** defines the name of the storage class to use 
          - **resources** defines how much storage we want (1 gigabyte).

4. Tell Kubernetes to create the StatefulSet from the manifest we just created with the following command:

    ```
    kubectl create -f ss.yaml
    ```

    A message displays to confirm the creation of the StatefulSet.

## Step 2: Check creation of Block Storage volumes

1. Run the following command in the terminal to check the progress of creation of your StatefulSet:

    ```
    kubectl get all
    ```

    It may take a few minutes for all 5 replica pods to be created, but when this process has finished you should see an output similar to the following:

    ```
    NAME                READY   STATUS    RESTARTS    AGE
    pod/ss-csi-scw-0    1/1     Running   0           63s
    pod/ss-csi-scw-1    1/1     Running   0           57s
    pod/ss-csi-scw-2    1/1     Running   0           42s
    pod/ss-csi-scw-3    1/1     Running   0           32s
    pod/ss-csi-scw-4    1/1     Running   0           22s

    NAME                          READY   AGE
    statefulset.apps/ss-csi-scw   5/5     63s
    ```

>     Note that each pod has it's own identity.

Reviewer: ```suggestion
    Note that each pod has its own identity.
```
